src/MessageTrait.php

```diff
@@ -170,8 +178,11 @@ public function withHeader($header, $value)
             );
         }
 
+        $lowerHeader = strtolower($header);
+
         $new = clone $this;
-        $new->headers[$header] = $value;
+        $new->headerNames[$lowerHeader] = $header;
+        $new->headers[$lowerHeader] = $value;
```

Comment:
> One argument brought up on the FIG mailing list is that the headers should be returned using the same case as was used to register them.

If we should respect this condition, then your/my implementation failed...

Should I introduce a mapping between lower cased header name and a collection of case sensitive header names and merge everything when we ask for a specific header?

Headers names would result to:

``` php
$names = [
    'x-foo' => ['X-Foo', 'X-foo'],
];
```

and the headers would not be affected (keeping original input headers).

What do you think?